platforms/software/ivy/src/main/java/org/gradle/api/publish/ivy/internal/publication/IvyComponentParser.java

```diff
@@ -403,4 +402,39 @@ public ParsedVariantDependencyResult(
             this.warnings = warnings;
         }
     }
+
+    private static class ArtifactKey {
+
+        private final File file;
```

Comment:
By adding file to the key, we will now have the possibility of publishing more than one artifact with the same "id" (name, ext, type, classifier). Before, if someone tried to publish two artifacts with the same ID, we'd just ignore the second. I'm not sure this is compliant with the Ivy spec, but it seems odd to do

I'd prefer to stick with the old behavior. The best behavior would be to (deprecate and) make it an error to try to do this, but I think this change is out of the scope of this PR

Reply:
Which integ test was failing?

Reply:
[IvyPublishFeaturesJavaIntegTest](https://github.com/gradle/gradle/blob/0bdd7c38e3245bc7201e80535650a86cbb3159d6/platforms/software/ivy/src/integTest/groovy/org/gradle/api/publish/ivy/IvyPublishFeaturesJavaIntegTest.groovy#L167)

Reply:
And what behavior changed that actually warranted adding the file to this key?

Reply:
IIRC, it's the diff between `""` and `null` as classifier

Reply:
I don't see how that change makes it so we should include the file in the key

Reply:
Here, in master:
* https://github.com/gradle/gradle/blob/ebfe7564f2c8ef7bf1a502b4ab4a45cb269be2d1/platforms/software/ivy/src/main/java/org/gradle/api/publish/ivy/internal/publication/IvyComponentParser.java#L162

The artifact key is calculated based on classifier, among other things. In the test, two artifacts are published: one with `""` classifier, another with `null` classifier. They also have different files. 
So, when I normalized their classifier, their classifier became the same, leading to the same key, which resulted in test failure. Thus I included the file to the artifact key